smithy-cli/src/main/java/software/amazon/smithy/cli/commands/UpgradeCommand.java

```diff
@@ -0,0 +1,522 @@
+                            LOGGER.finest("Skipping non-target model file: " + locationPath);
+                            return false;
+                        }
+                    }
+                    if (!locationPath.toString().endsWith(".smithy")) {
+                        LOGGER.info("Skipping non-IDL model file: " + locationPath);
+                        return false;
+                    }
+                    return true;
+                })
+                .sorted()
+                .collect(Collectors.toList());
+    }
+
+    private void writeUpgradedFile(Model completeModel, Path filePath) {
+        try {
+            Files.write(filePath, upgradeFile(completeModel, filePath).getBytes(StandardCharsets.UTF_8));
+        } catch (IOException e) {
+            throw new CliError(format("Unable to write upgraded model file to %s: %s", filePath, e));
+        }
+    }
+
+    String upgradeFile(Model completeModel, Path filePath) {
+        String contents = IoUtils.readUtf8File(filePath);
+        if (VERSION_2.matcher(contents).find()) {
+            return contents;
+        }
+
+        ShapeUpgradeVisitor visitor = new ShapeUpgradeVisitor(completeModel, contents);
+
+        completeModel.shapes()
+                .filter(shape -> shape.getSourceLocation().getFilename().equals(filePath.toString()))
+                // Apply updates to the shapes at the bottom of the file first.
+                // This lets us modify the file without invalidating the existing
+                // source locations.
+                .sorted(new SourceLocationSorter().reversed())
+                .forEach(shape -> shape.accept(visitor));
+
+        return updateVersion(visitor.getModelString());
+    }
+
+    private String updateVersion(String modelString) {
+        Matcher matcher = VERSION_1.matcher(modelString);
+        if (matcher.find()) {
+            return matcher.replaceFirst(format("\\$version: \"2.0\"%n"));
+        }
+        return format("$version: \"2.0\"%n%n") + modelString;
+    }
+
+    private static final class ResultConsumer implements Consumer<ProjectionResult>, BiConsumer<String, Throwable> {
+        private Throwable error;
+        private ProjectionResult result;
+
+        @Override
+        public void accept(String name, Throwable throwable) {
+            this.error = throwable;
+        }
+
+        @Override
+        public void accept(ProjectionResult projectionResult) {
+            // We only expect one result because we're only building one
+            // projection - the source projection.
+            this.result = projectionResult;
+        }
+
+        ProjectionResult getResult() {
+            if (error != null) {
+                throw new RuntimeException(error);
+            }
+            return result;
+        }
+    }
+
+    // Sorts shapes in the order they appear in the file.
+    private static class SourceLocationSorter implements Comparator<FromSourceLocation>, Serializable {
```

Comment:
This can be replaced with the Sorter added in This can be replaced with the Sorter added in https://github.com/awslabs/smithy/pull/1192